Zend/zend_object_handlers.c

```diff
@@ -1334,15 +1321,21 @@ ZEND_API void zend_std_unset_property(zend_object *zobj, zend_string *name, void
 	if (EXPECTED(IS_VALID_PROPERTY_OFFSET(property_offset))) {
 		zval *slot = OBJ_PROP(zobj, property_offset);
 
-		if (Z_TYPE_P(slot) != IS_UNDEF) {
-			if (UNEXPECTED(prop_info && (prop_info->flags & ZEND_ACC_READONLY))) {
-				if (Z_PROP_FLAG_P(slot) & IS_PROP_REINITABLE) {
-					Z_PROP_FLAG_P(slot) &= ~IS_PROP_REINITABLE;
-				} else {
-					zend_readonly_property_unset_error(prop_info->ce, name);
-					return;
-				}
+		if (prop_info && UNEXPECTED(prop_info->flags & (ZEND_ACC_READONLY|ZEND_ACC_PPP_SET_MASK))
+		 && (Z_TYPE_P(slot) != IS_UNDEF || (Z_PROP_FLAG_P(slot) & IS_PROP_UNINIT) || !zobj->ce->__unset)) {
```

Comment:
The `!zobj->ce->__unset` check is not covered by tests currently:

``` php
class C {
    public protected(set) int $a;
    public function __construct() {
        unset($this->a);
    }
}

$c = new C();
unset($c->a);
```

While testing, I found that recursive unsets in __unset() do not check asymmetric visibility in some case:

``` php
class C {
    public private(set) int $a = 1;
    public function __construct() {
        unset($this->a);
    }
}

class D extends C {
    public function __unset($name) {
        unset($this->a);
    }
}

$c = new D();
unset($c->a);
```

This is also the case for writes:

``` php
class C {
    public private(set) int $a = 1;
    public function __construct() {
        unset($this->a);
    }
}

class D extends C {
    public function __set($name, $value) {
        $this->a = $value;
    }
}

$c = new D();
$c->a = 2;
var_dump($c);
```

Reply:
Good catch. Unfortunately, we need to check for the property guard.